lib/websocket/connection.js

```diff
@@ -0,0 +1,365 @@
+  ws[kReadyState] = states.OPEN
+
+  // 2. Change the extensions attribute’s value to the extensions in use, if
+  //    it is not the null value.
+  // https://datatracker.ietf.org/doc/html/rfc6455#section-9.1
+  const extensions = response.headersList.get('sec-websocket-extensions')
+
+  if (extensions !== null) {
+    ws[kExtensions] = extensions
+  }
+
+  // 3. Change the protocol attribute’s value to the subprotocol in use, if
+  //    it is not the null value.
+  // https://datatracker.ietf.org/doc/html/rfc6455#section-1.9
+  const protocol = response.headersList.get('sec-websocket-protocol')
+
+  if (protocol !== null) {
+    ws[kProtocol] = protocol
+  }
+
+  // 4. Fire an event named open at the WebSocket object.
+  fireEvent('open', ws)
+}
+
+/**
+ * @see https://websockets.spec.whatwg.org/#feedback-from-the-protocol
+ * @param {import('./websocket').WebSocket} ws
+ */
+function receiveData (ws) {
+  const { [kResponse]: response } = ws
+
+  /** @type {WebsocketFrame|undefined} */
+  let frame
+
+  response.socket.on('data', (chunk) => {
+    if (!frame) {
+      frame = WebsocketFrame.from(chunk)
+      frame.terminated = frame.fin // message complete
+    } else {
+      // A fragmented message consists of a single frame with the FIN bit
+      // clear and an opcode other than 0, followed by zero or more frames
+      // with the FIN bit clear and the opcode set to 0, and terminated by
+      // a single frame with the FIN bit set and an opcode of 0.
+
+      const fragmentedFrame = WebsocketFrame.from(chunk)
+
+      if (fragmentedFrame.fin && fragmentedFrame.opcode === opcodes.CONTINUATION) {
+        frame.terminated = true
+      }
+
+      frame.addFrame(fragmentedFrame.data)
+    }
+
+    if (frame.opcode === opcodes.CLOSE) {
+      // Upon either sending or receiving a Close control frame, it is said
+      // that _The WebSocket Closing Handshake is Started_ and that the
+      // WebSocket connection is in the CLOSING state.
+      ws[kReadyState] = states.CLOSING
+      ws[kClosingFrame] = frame
+      frame = undefined
+      return
+    }
+
+    // rsv bits are reserved for future use; if any aren't 0,
+    // we currently can't handle them.
+    if (frame.rsv1 || frame.rsv2 || frame.rsv3) {
+      failWebsocketConnection(ws[kController])
+      return
+    }
+
+    // If the frame comprises an unfragmented
+    // message (Section 5.4), it is said that _A WebSocket Message Has Been
+    // Received_
+
+    if (frame.terminated) {
+      // 1. If ready state is not OPEN (1), then return.
+      if (ws[kReadyState] !== states.OPEN) {
+        return
+      }
+
+      // 2. Let dataForEvent be determined by switching on type and binary type:
+      let dataForEvent
+
+      if (frame.opcode === opcodes.TEXT) {
+        // - type indicates that the data is Text
+        //      a new DOMString containing data
+        dataForEvent = new TextDecoder().decode(frame.data)
+      } else if (frame.opcode === opcodes.BINARY) {
```

Comment:
You can also receive a control (ping/pong) frame. You can ignore them but you don't want to return them as messages via the `'message'` event to the user.

Also you should handle pings. A peer must respond with a pong when it receives a ping. Yes, browsers also do this.